### PR TITLE
api: validate snaps against validation set assert from the store

### PR DESCRIPTION
**Pull request by stolowski**
Handle the case of `snap validate <..>' against store (latest or specific sequence number) when we don't have the given set locally.

### CODE REVIEW

File: daemon/api_validate.go [hunk 1]
```diff
@@ -141,7 +143,29 @@ func listValidationSets(c *Command, r *http.Request, _ *auth.UserState) Response
 	return SyncResponse(results, nil)
 }
 
-func getValidationSet(c *Command, r *http.Request, _ *auth.UserState) Response {
+var checkInstalledSnaps = func(vsets *snapasserts.ValidationSets, snaps []*snapasserts.InstalledSnap) error {
+	return vsets.CheckInstalledSnaps(snaps)
```
Comment (pedronis) [inline daemon/api_validate.go:147]: this is never tested, also given it has no side-effects , I suppose we could call it wrap it to do extra checks but still call it for real as well?  

we use that pattern of wrap but call in some other places in the daemon tests I think

File: daemon/api_validate.go [hunk 2]
```diff
@@ -289,3 +313,55 @@ func forgetValidationSet(st *state.State, accountID, name string, sequence int)
...
+	if at == nil {
+		return nil, fmt.Errorf("internal error: validation-set assert type not found")
+	}
+
+	model, err := c.d.overlord.DeviceManager().Model()
+	if err != nil {
+		return nil, err
+	}
+
+	sequenceKey := []string{model.Series(), accountID, name}
```
Comment (pedronis) [inline daemon/api_validate.go:360]: we just use usually release.Series and not consult the actual model

File: daemon/api_validate.go [hunk 3]
```diff
@@ -289,3 +313,55 @@ func forgetValidationSet(st *state.State, accountID, name string, sequence int)
...
+		Valid: validErr == nil,
+	}
+	return SyncResponse(res, nil)
+}
+
+func getSingleSeqFormingAssertion(c *Command, accountID, name string, sequence int, user *auth.UserState) (asserts.Assertion, error) {
+	sto := snapstate.Store(c.d.overlord.State(), nil)
+	at := asserts.Type("validation-set")
+	if at == nil {
+		return nil, fmt.Errorf("internal error: validation-set assert type not found")
```
Comment (pedronis) [inline daemon/api_validate.go:352]: this could probably just panic

File: daemon/api_validate.go [hunk 4]
```diff
@@ -289,3 +313,55 @@ func forgetValidationSet(st *state.State, accountID, name string, sequence int)
...
+		return InternalError(err.Error())
+	}
+
+	validErr := checkInstalledSnaps(sets, snaps)
+	res := validationSetResult{
+		AccountID: vset.AccountID(),
+		Name:      vset.Name(),
+		Sequence:  vset.Sequence(),
+		// XXX: pass actual err details and implement "verbose" mode
+		// for the client?
```
Comment (pedronis) [inline daemon/api_validate.go:342]: let's leave a TODO here and consider it again once we have implemented a bit more of the rest

**Review by pedronis — COMMENTED** [review]
did a first pass, looking good but some comments

File: daemon/api_validate.go (hunk 1 above)
Comment (stolowski) [reply daemon/api_validate.go:148]: Done.

**Review by pedronis — APPROVED** [review]
thank you

File: daemon/api_validate.go [hunk 5]
```diff
@@ -141,7 +144,29 @@ func listValidationSets(c *Command, r *http.Request, _ *auth.UserState) Response
...
+		return nil, err
+	}
+	for _, snapState := range all {
+		cur, err := snapState.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		snaps = append(snaps, snapasserts.NewInstalledSnap(
+			snapState.InstanceName(), snapState.CurrentSideInfo().SnapID,
+			cur.Revision))
```
Comment (bboozzoo) [inline daemon/api_validate.go:164]: nitpicking, indentation is slightly confusing here, feel free to ignore
```suggestion
		snaps = append(snaps,
			snapasserts.NewInstalledSnap(snapState.InstanceName(),
				snapState.CurrentSideInfo().SnapID,
				cur.Revision))
```

**Review by bboozzoo — APPROVED** [review]
LGTM

**Pull request merged**